[FSDP] Add device_id argument

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75977
* #75976
* #75970
* #75945

device_id argument will move module to CUDA device specified by ID, to
allow sharding, broadcast, flattening etc to take place on CUDA device even for
passed in CPU modules to help the initialization time.

Similar to DDP, inputs are also cast if this argument is specified. `to_kwargs`
from DDP is generalized.

Differential Revision: [D35712024](https://our.internmc.facebook.com/intern/diff/D35712024/)

rohan-varma committed:
[FSDP] Add device_id argument

device_id argument will move module to CUDA device specified by ID, to
allow sharding, broadcast, flattening etc to take place on CUDA device even for
passed in CPU modules to help the initialization time.

Similar to DDP, inputs are also cast if this argument is specified. `to_kwargs`
from DDP is generalized.

Differential Revision: [D35712024](https://our.internmc.facebook.com/intern/diff/D35712024/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -373,6 +374,11 @@ class FullyShardedDataParallel(nn.Module):
             avoid sharding specific parameters when using an
             ``auto_wrap_policy`` or if parameters' sharding is not managed by
             FSDP. (Default: ``None``)
+        device_id (Optional[int]): An `int` describing the CUDA device FSDP
```
nit: I think in the docstring, we need double backtick to render as code, so \'int\' -> \`\`int\`\`.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -373,6 +374,11 @@ class FullyShardedDataParallel(nn.Module):
             ``auto_wrap_policy`` or if parameters' sharding is not managed by
             FSDP. (Default: ``None``)
+        device_id (Optional[int]): An `int` describing the CUDA device FSDP
+            module should be moved to before initialization such as sharding
```
nit: A bit confused on the comment. Maybe it should be `module should be moved to before initialization when sharding takes place`?

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1571,6 +1585,12 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
         self.training_state = TrainingState_.FORWARD
 
+        if self._is_root:
+            if self.device_id:
```
Should we change to this `if self.device_id is not None`? It might confound `self.device_id == 0` and `self.device_id is None` right now.

awgu commented on torch/distributed/utils.py:
```diff
@@ -1,8 +1,92 @@
     return dist._verify_params_across_processes(process_group, tensors, logger)
 
+
+def _recursive_to(inputs, target_gpu, use_side_stream_for_tensor_copies):
```
nit: Adding type annotations might be helpful. For example to clarify that `target_gpu` is an `int` and not a `torch.device` it seems?

awgu commented on torch/distributed/utils.py:
```diff
@@ -1,8 +1,92 @@
+    return res
+
+
+def _to_kwargs(inputs, kwargs, device_id, use_side_stream_for_tensor_copies):
```
nit: Similar comment here. Maybe good to add some type annotations.

rohan-varma replied:
I agree we should add type annotations, although for this PR I'd like to just move the code as is. I'll file an issue to add the type annotations as a follow up.

awgu commented on torch/nn/parallel/distributed.py:
```diff
@@ -36,8 +37,7 @@
 from ..modules import Module
-from ._functions import _get_stream
-from .scatter_gather import gather, is_namedtuple, scatter_kwargs
+from .scatter_gather import gather, scatter_kwargs
```
I may have missed it, but where is `gather()` used?

rohan-varma replied:
https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L1135, it is not used by DDP, but I guess it is a public interface, so we cannot easily remove it

awgu reviewed (commented):
Left some initial comments.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -431,6 +438,13 @@ def __init__(
+        if self.device_id is not None:
+            # Move module to device specified. Note that this is done prior to
+            # setting compute_device to ensure that they align.
+            module = module.to(self.device_id)
```
check: if module is a GPU module, the GPU module's device is the same as device_id, and also similar to DDP, check the GPU module was placed in the single device

rohan-varma replied:
Sure, will add this. But this is different from DDP right, DDP supports multi-device modules (it is how DDP(Pipe()) works?)

mrshenli replied:
> DDP supports multi-device modules (it is how DDP(Pipe()) works?)

Yep. I am wondering if FSDP might also need to support similar multi-device in the future (low-pri for now)? At least that could work with FSDP without param sharding?

rohan-varma replied:
For now we'll assume multi-device support is not there because we haven't actively tested / worked on this?

zhaojuanmao reviewed (commented):
this approach looks good to me, also DDP users possibly feel it is familiar and feel comfortable to use it.

rohan-varma committed:
Update on "[FSDP] Add device_id argument"


device_id argument will move module to CUDA device specified by ID, to
allow sharding, broadcast, flattening etc to take place on CUDA device even for
passed in CPU modules to help the initialization time.

Similar to DDP, inputs are also cast if this argument is specified. `to_kwargs`
from DDP is generalized.

Differential Revision: [D35712024](https://our.internmc.facebook.com/intern/diff/D35712024/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -376,6 +377,11 @@ class FullyShardedDataParallel(nn.Module):
             avoid sharding specific parameters when using an
             ``auto_wrap_policy`` or if parameters' sharding is not managed by
             FSDP. (Default: ``None``)
+        device_id (Optional[int]): An `int` describing the CUDA device FSDP
```
nit: do we need to stay consistent with DDP? DDP's doc allows both `int` and `torch.device`. 

https://github.com/pytorch/pytorch/blob/6ca8272d46a67b56b1cb65bb551bcbf5709f005b/torch/nn/parallel/distributed.py#L428


BTW, Future's device allows int, str, and device:

https://github.com/pytorch/pytorch/blob/6ca8272d46a67b56b1cb65bb551bcbf5709f005b/torch/futures/__init__.py#L22

rohan-varma replied:
Yep, it should be int or torch.device. Will fix

rohan-varma commented:
Still need to add unittests to the PR.

rohan-varma committed:
Update on "[FSDP] Add device_id argument"


device_id argument will move module to CUDA device specified by ID, to
allow sharding, broadcast, flattening etc to take place on CUDA device even for
passed in CPU modules to help the initialization time.

Similar to DDP, inputs are also cast if this argument is specified. `to_kwargs`
from DDP is generalized.

Differential Revision: [D35712024](https://our.internmc.facebook.com/intern/diff/D35712024/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Add device_id argument"


device_id argument will move module to CUDA device specified by ID, to
allow sharding, broadcast, flattening etc to take place on CUDA device even for
passed in CPU modules to help the initialization time.

Similar to DDP, inputs are also cast if this argument is specified. `to_kwargs`
from DDP is generalized.

Differential Revision: [D35712024](https://our.internmc.facebook.com/intern/diff/D35712024/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Add device_id argument"


device_id argument will move module to CUDA device specified by ID, to
allow sharding, broadcast, flattening etc to take place on CUDA device even for
passed in CPU modules to help the initialization time.

Similar to DDP, inputs are also cast if this argument is specified. `to_kwargs`
from DDP is generalized.

Differential Revision: [D35712024](https://our.internmc.facebook.com/intern/diff/D35712024/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Add device_id argument"


device_id argument will move module to CUDA device specified by ID, to
allow sharding, broadcast, flattening etc to take place on CUDA device even for
passed in CPU modules to help the initialization time.

Similar to DDP, inputs are also cast if this argument is specified. `to_kwargs`
from DDP is generalized.

Differential Revision: [D35712024](https://our.internmc.facebook.com/intern/diff/D35712024/)

[ghstack-poisoned]

rohan-varma commented:
https://github.com/pytorch/pytorch/pull/77321
